src/transform/src/canonicalization/flatmap_to_map.rs

```diff
@@ -42,70 +43,59 @@ impl crate::Transform for FlatMapElimination {
 }
 
 impl FlatMapElimination {
-    /// Turns `FlatMap` into `Map` if only one row is produced by flatmap.
+    /// See comment at the top of the file.
     pub fn action(relation: &mut MirRelationExpr) {
         if let MirRelationExpr::FlatMap { func, exprs, input } = relation {
-            let (func, with_ordinality) = if let TableFunc::WithOrdinality { inner } = func {
-                // get to the actual function, but remember that we have a WITH ORDINALITY clause.
-                (&**inner, true)
-            } else {
-                (&*func, false)
-            };
-
-            if let TableFunc::GuardSubquerySize { .. } = func {
-                // (`with_ordinality` doesn't matter because this function never emits rows)
-                if let Some(1) = exprs[0].as_literal_int64() {
-                    relation.take_safely(None);
-                }
-            } else if let TableFunc::Wrap { width, .. } = func {
-                if *width >= exprs.len() {
-                    *relation = input.take_dangerous().map(std::mem::take(exprs));
-                    if with_ordinality {
-                        *relation = relation.take_dangerous().map_one(MirScalarExpr::literal(
-                            Ok(Datum::Int64(1)),
-                            ScalarType::Int64,
-                        ));
-                    }
-                }
-            } else if is_supported_unnest(func) {
-                let func = func.clone();
-                let exprs = exprs.clone();
-                use mz_expr::MirScalarExpr;
-                use mz_repr::RowArena;
-                if let MirScalarExpr::Literal(Ok(row), ..) = &exprs[0] {
-                    let temp_storage = RowArena::default();
-                    if let Ok(mut iter) = func.eval(&[row.iter().next().unwrap()], &temp_storage) {
-                        match (iter.next(), iter.next()) {
-                            (None, _) => {
-                                // If there are no elements in the literal argument, no output.
+            let args: Result<Vec<Datum>, ()> = exprs
+                .iter()
+                .map(|e| {
+                    e.as_literal()
+                        // give up if one of them is not a literal
+                        .ok_or(())?
+                        // give up also if one of them is a literal error
+                        .map_err(|_eval_err| ())
+                })
+                .collect();
```

Comment:
nit: the short-circuiting behavior of collecting a Result is a big magical, maybe the imperative version reads clearer

```rust
let args = vec![];
for expr in exprs {
    match expr.as_literal() {
        Some(Ok(datum)) => args.push(datum),
        _ => return;
    }
}
```

Reply:
Hmm, ok! I thought I'm the only one who finds this hard! But seeing that a real Rust expert also doesn't like it, I'll definitely switch to the imperative version.